### PR TITLE
Custom stacks

### PR DESCRIPTION
**Pull request by raedwulf**
Alternative implementation - which makes stack allocators return the top of the stack (thus the management of page guards makes more sense with the stack allocator).

Performance differences are negligible.  This implementation also supports page guards within the stack pool as opposed to #32 .

### CODE REVIEW

**Comment by sustrik** [comment]
This patch is too big to review in one go. I'll start pulling it in piece by piece. I guess I'll start with alloca-based stack switching. You are saying that it's more robust than VLAs. What's the case when VLAs break and alloca works? Is it the stack protector on modern gccs?

**Comment by raedwulf** [comment]
I half-answered it in the other pull-request, but in summary for all the key points:

- `alloca` works when there's no proper VLA support and doesn't require two sets of `volatile`.
- It also has the advantage of being used in pre-C99 code for setting the stack, so there's a better chance that `alloca` behaviour will remain the same for future compiler versions for backwards-compatibility.
- `clang` does not accept zero-sized arguments to `alloca`, so in that case I do allocate a little, but `gcc` is happy to accept zero-sized `alloca` (as it does for arrays too).
- GCC also treats `alloca` partially like a function call in the generated assembly and appears to expect the stack to be clobbered whereas VLAs are subject to more optimisation by the compiler.

**Comment by sustrik** [comment]
Ok, I've split the alloca stuff. Along the way I've simplified it based on assumption that alloca is always available (it's a built-in so I guess it should always be the case with gcc/clang).

hquery optimisation: https://github.com/sustrik/libdill/commit/286a04bbd503ef8478e84c0bd11a1d351dc08fda

alloca for x86-64 and i386:
https://github.com/sustrik/libdill/commit/c5b1e10e8370cc7e21a682ce978d818475cac307

alloca for arch fallback:
https://github.com/sustrik/libdill/commit/061d8f2f7fcd1ca91a2fe712d1a10b9e6ad7e3dc

I would appreciate a code review.

**Comment by sustrik** [comment]
Merged.

Now I am looking at allocators. Can you give more context about this (apool):

>   The reason for the separation is to map neatly to an underlying allocator
>   which has fixed size allocation without significantly wasting space per
>   allocation.

**Comment by sustrik** [comment]
Generally speaking, there are 2 use cases for allocators:

1. User doesn't mind. They use go() and want it to work without any extra work. We should make effort to make this work as smoothly, safely and efficiently as possible.
2. Embedded developer. The device is short on memory. It doesn't have memory overcommit. The developer wants to handle stack sizes by hand (measure using census, then fill in a concrete number).

There is a third pseudo-use-case to guarantee system's generality:

3. Custom allocator written by the user. The requirement is that built-in allocators should not have extra status. It should be possible to re-implement any  built-in allocator as a custom allocator.

**Comment by sustrik** [comment]
Another interesting use case: Embedded developer wants to use a library, say dsock. The library uses default-allocated stacks which causes the program to run out of memory on the embedded device. What now?

**Comment by raedwulf** [comment]
> The reason for the separation is to map neatly to an underlying allocator
> which has fixed size allocation without significantly wasting space per
> allocation.

I added some comments to the last commit: 2d5e9ed0fe62b5287027b35f3b66d461ae577d4d

* Currently the pool internally uses a single `blocksize` (element size * element count bytes).  It currently allocates 2 blocks (initially) of that `blocksize`  
* There's the pool block where each element size is the exact size the user wants to allocate.
* There's the block header which contains links to each pool block present.
* When one pool block is exhausted, another pool block will get allocated.
* If the block header is exhausted, another block header is allocated and another pool block is allocated as well.
* If I didn't make the separation of the block header and the pool block, additional pool blocks need to contain a link to the next pool blocks which would take a little of the blocksize which the user might want aligned to a page.
  - But! 4096 - 8 = 4088... which doesn't divide cleanly
  - 4096 + 8 = 4114 -> 8192... which is two pages.
  - My solution was to contain the 8 bytes for each pool block link in another block.
* I should separate the actual block header size and the pool block size later it later on so that there's finer grained control.
* I had this functionality in my original implementation (which I wrote for another one of my projects).  However, we could simplify it if we drop block extensions and just return ENOMEM when a single pool is exhausted.

Feature differences from the cached allocator:

- Pooled allocator keeps growing (but that could be changed to a simpler fixed size model)
- Pooled allocator tracks all memory allocated (cache only tracks up to 64)
- Pooled allocator is slightly faster (might be due to the implementation of free lists)
- Page guards are supported in pools like caches but unlike NOGUARD mallocs (if you have some custom malloc hook)

My use case for the pool allocator was that you could have a 100,000 to a million coroutines on a server: C100K and C1000K problems.  Each connection having its own coroutine.  The pool allocator makes it possible to track every one of these coroutine stacks whereas the cache can't.

You could setthreads of the minimum size = 4096 bytes and that would reach about 42 gb virtual memory if you have 1 million coroutines.  Since it's virtual memory, I managed to achieve this with some swapping on 32 gb RAM.

This could be used through the `go_stack` interface as well.

Of course you probably run into other problems before reaching 1 million connections, but at least libdill coroutines wouldn't be the reason for it :P

**Comment by sustrik** [comment]
The reason why I am asking is that if this allocator is superior to the default one we have now, why not use it as a default?

That being said I am still not sure about what it tries to do:

* Pooled allocator keeps growing (but that could be changed to a simpler fixed size model)
  - with default allocator the cache is fixed (64 stacks) but the total amount of stacks is unlimited
* Pooled allocator tracks all memory allocated (cache only tracks up to 64)
  - why do we need to track stacks in use? it's not like the memory becomes unreachable. it's reachable via coroutine handle.
* Pooled allocator is slightly faster (might be due to the implementation of free lists)
  - fair enough

Maybe it is that you are trying to optimise memory layout of stacks by avoiding malloc's chunk header?

**Comment by raedwulf** [comment]
> Another interesting use case: Embedded developer wants to use a library, say dsock. The library uses default-allocated stacks which causes the program to run out of memory on the embedded device. What now?

This is part of the ownership problem I've been struggling with.  My initial implementation could set the default allocator of libdill so that any subsequent `go` calls would use the default allocator set by the application.
In this case, the default allocator could have been implemented by the embedded developer so that they could track allocations more efficiently in a custom manner.

This does raise the question about `go_stack`.  How do you know if a coroutine handle `h` has been allocated using the default allocator or a custom allocator.  `libdill` does track this internally, but doesn't expose an API for it.  This could be an argument for the earlier API where you passed an allocator handle or an alternative `destructor` function (although this breaks the API model).

**Comment by raedwulf** [comment]
> Pooled allocator keeps growing (but that could be changed to a simpler fixed size model)
with default allocator the cache is fixed (64 stacks) but the total amount of stacks is unlimited

The limit of 64-stacks means that you can hit >64-allocation/free cycles.  Admittedly, there's no application that uses libdill yet, so whether this will be a true problem remains to be seen.

> Pooled allocator tracks all memory allocated (cache only tracks up to 64)
why do we need to track stacks in use? it's not like the memory becomes unreachable. it's reachable via coroutine handle.

It's more useful, than necessary, it could potentially let us/the developer profile the stack pool usage.  It also allows us to own the memory of the stacks, and release all the stack memory at the end.  However, I admit I haven't fully succeeded in making valgrind happy because I think the stack switching confuses it.

> Maybe it is that you are trying to optimise memory layout of stacks by avoiding malloc's chunk header?

Yep.

> The reason why I am asking is that if this allocator is superior to the default one we have now, why not use it as a default?

It is possible, but it is more complex and makes debugging more difficult.  Caches have the advantage of simplicity - which helps valgrind understand what is going on.  Valgrind churns out a tonne of warnings with the pool implementation.
I'm reaching quite a high-level of confidence that my pool implementation works correctly - I tested my original implementation well.  I think I have now ironed out the bugs with some of my libdill-specific modifications that make it work better with pages.

**Comment by sustrik** [comment]
Ok, fair enough. Let's keep the current default allocator but think of a way to plug in custom allocators cleanly.

As for the ownership problem, my feeling is that there's no way to solve it easily. Either dsock in embedded-friendly, in which case the user doesn't have to care or it is not, in which case it can't be used in enbedded environments.

As for the plugin mechanism, my idea was that go_stack() is the low level mechanism. The handle represents the coroutine itself, but not the stack it is running on. go() is built on top of it, wrapping the underlying go_stack() handle, adding the allocation and creating a new handle to represent both coroutine and the stack. Question: how much will creating 2 handles per go() slow it down?

Marginally related: I would like to overload go_stack() to use default allocator when census is turned on. After coroutine finishes census would look at coroutine's stack usage and compare that with length passed to go_stack(). If will assert if it's using more stack than it should. That will allow the user to run tests in such a way that they fail if assumptions about stack sizes are violated.

**Comment by raedwulf** [comment]
`how much will creating 2 handles per go() slow it down?`

The times I'll use are based on my system using `perf/go`.  Current head runs at around 26ns per creation/termination.

I suppose it depends on the mechanism for allocating/creating the structure associated with the second handle.  At the moment, the handle for `go` is cheap because it allocates the handle using the stack allocator and the handle effectively has a form of memory pool.

I found accessing an allocator (which has it's own vfs + handle) slows down the overall goroutine slightly.
On my earlier implementation (diverged from HEAD now), https://github.com/raedwulf/libdill/tree/optimized-limit
was getting 31-32ns (~20% slower) with a heavily optimised default allocator fast path.  Without the fast path, default allocators were taking around 38ns (~50% slower).

**Comment by raedwulf** [comment]
> Marginally related: I would like to overload go_stack() to use default allocator when census is turned on. After coroutine finishes census would look at coroutine's stack usage and compare that with length passed to go_stack(). If will assert if it's using more stack than it should. That will allow the user to run tests in such a way that they fail if assumptions about stack sizes are violated.

I guess in this instance, `#ifdef DILL_CENSUS` over `#define go_stack` in `libdill.h` and provide two different implementations would work.

To add to my above comment on pluggable allocators. I think we're running into the issue where abstraction will always run into performance issues unless you take C++'s approach of making everything a meta-programmed header nightmare (which is possible with C macros as well).

Another alternative would be to provide separate builds with different allocators.  The advantage of this would be that the allocators are compatible and just requires a linking switch.
The downside is that the allocators have to be compatible otherwise an application will run the danger of assuming a stack size too large or too small.  So perhaps what we need to support is just to be able to set the default allocator size.

```C
size_t stksize();
int setstksize(size_t size); /* returns 0 on success; returns -1 on failure and sets errno=EINVAL/ENOMEM depending on certain conditions.  EINVAL for perhaps bigger than allocator support and ENOMEM for just out of memory due to block size being too big (e.g. pool) */
```

Potentially we could support `int stkguard()` and `int setstkguard(int enable)` as well.

This would fit well in the current model of enabling/disabling features of libdill.  Perhaps making it possible to build multiple versions with the build scripts would be handy for maintainers.
In summary:
```bash
libdill.so # Standard libdill library (--disable-census, --disable-arch-fallback, --disable-pool-allocator, --enable-guard)
libdill-census.so # (--enable-census, --disable-arch-fallback, --disable-pool-allocator, --enable-guard)
libdill-pool.so # (--disable-census, --disable-arch-fallback, --enable-pool-allocator, --disable-guard)
libdill-pool-census.so # (--enable-census, --disable-arch-fallback, --enable-pool-allocator, --disable-guard)
```

**Comment by raedwulf** [comment]
If someone develops their own allocator, they can test it using `go_stack` first and we can ask them to submit a pull-request if it has general-relevance to developing with libdill and is compatible with the existing allocators.

**Comment by sustrik** [comment]
Ok, what about getting rid of the allocator handles altogether? (Btw, it's also needed to allow for global, statically defined allocator instances -- handles can't be statically defined.)

```
struct stackalloc_vfs {
    void *(*alloc)(struct stackalloc_vfs *vfs, size_t len);
    int (*free)(struct stackalloc_vfs *vfs, void *ptr, size_t len);
};

DILL_EXPORT extern const struct stackalloc_vfs *stackalloc_default;

#define go_stack(fn, ptr, len) ...
#define go_alloc(fn, alloc, len) ...
#define go(fn) go_alloc(fn, stackalloc_default, 256 * 1024)
```

Given that the vfs pointer is const, the compiler may be even able to inline the function (not sure whether it actually will).

**Comment by raedwulf** [comment]
That could work.  The compiler can't inline this because the symbols within the application will not get resolved until linking.  There might be a chance that the compiler can optimise if libdill is completely embedded in an application and whole program optimisation is used but I don't think it works with function parameters currently.

**Comment by raedwulf** [comment]
Performance-wise that should be a lot faster ~ if I have some time later I'll make those changes.

**Comment by raedwulf** [comment]
Closing these as master has diverged very far from these branches, so a rebase would be needed.
